mlrun/package/packager.py

```diff
@@ -0,0 +1,537 @@
+      the ``DataItem`` is "object" then the class method ``unpack_object`` must be implemented. The signature of each
+      unpack class method must be:
+
+      >>> @classmethod
+      ... def unpack_x(cls, data_item: mlrun.DataItem, ...) -> Any:
+      ...     pass
+
+      Wehre 'x' is the artifact type, 'data_item' is the artifact's data item to unpack, ... are the instructitons that
+      were originaly returned from ``pack_x`` (Each instruction must be optional (have a default value) to support
+      objects from this type that were not packaged but customly logged) and the returning value is the unpacked
+      object.
+    * ``is_packable`` is getting the object and the artifact type desired to pack and log it as. So, it is automatically
+      looking for all pack class methods implemented to collect the supported artifact types. So, if ``PackagerX`` has
+      ``pack_y`` and ``pack_z`` that means the artifact typpes supported are 'y' and 'z'.
+
+    In order to link between packages (using the extra data or metrics spec attributes of an artifact), you should use
+    the key as if it exists and as value ellipses (...). The manager will link all packages once it is done packing. For
+    exmaple, given extra data keys in the log hint as `extra_data`, setting them to an artifact should be:
+
+    >>> artifact = Artifact(key="my_artifact")
+    >>> artifact.spec.extra_data = {key: ... for key in extra_data}
+    """
+
+    # The type of object this packager can pack and unpack:
+    TYPE: Type = ...
+    # The default artifact type to pack as or unpack from:
+    DEFAULT_ARTIFACT_TYPE = ArtifactTypes.OBJECT
+
+    @classmethod
+    def pack(
+        cls, obj: Any, artifact_type: Union[str, None], configurations: dict
+    ) -> Union[Tuple[Artifact, dict], dict]:
+        """
+        Pack an object as the given artifact type using the provided configurations.
+
+        :param obj:            The object to pack.
+        :param artifact_type:  Artifact type to log to MLRun.
+        :param configurations: Log hints configurations to pass to the packing method.
+
+        :return: If the packed object is an artifact, a tuple of the packed artifact and unpacking instructions
+                 dictionary. If the packed object is a result, a dictionary containing the result key and value.
+        """
+        # Get default artifact type in case it was not provided:
+        if artifact_type is None:
+            artifact_type = cls.DEFAULT_ARTIFACT_TYPE
+
+        # Get the packing method according to the artifact type:
+        pack_method = getattr(cls, f"pack_{artifact_type}")
+
+        # Validate correct configurations were passed:
+        cls._validate_method_arguments(
+            method=pack_method,
+            arguments=configurations,
+            arguments_type=cls._ArgumentsType.CONFIGURATIONS,
+        )
+
+        # Call the packing method and return:
+        return pack_method(obj, **configurations)
```

Comment:
I'm doing it because then every method the user will inherit will have to add either conf parameter or **kwargs but most importantly, to not fail the run if non matching arguments were provided, simply warn the user and go on.

Reply:
Passing a dictionary makes it easy - just take from the dict what parameters you're looking for, and warn if you get unrelated values, or don't get what you need. I agree that it's a bit easier to write extension packagers this way, since you just get parameters and don't need to unpack (pun intended) a dictionary. But, it's not like you really provide a lot more visibility or warnings by doing the kwarg expansion - for example, you don't warn on missing parameters, just if the parameters are different than the function expects.
It's eventually a style thing, I guess.

Reply:
I'll add better analysis for the parameters (missing required values and not only mismatching) or maybe @liranbg can add his opinion, because I still thinks its more convenient for the user to simply write a function and the rest is taken care of (passing them, validating them, etc).